src/asm/symbol.c

```diff
@@ -288,15 +295,47 @@ void sym_Purge(char const *symName)
 	} else if (isReferenced(sym)) {
 		error("Symbol \"%s\" is referenced and thus cannot be purged\n", symName);
 	} else {
-		// Do not keep a reference to the label's name after purging it
-		if (sym->name == labelScope)
-			sym_SetCurrentSymbolScope(NULL);
-
-		// FIXME: this leaks sym->macro for SYM_EQUS and SYM_MACRO, but this can't
-		// free(sym->macro) because the expansion may be purging itself.
-		hash_RemoveElement(symbols, sym->name);
-		// TODO: ideally, also unref the file stack nodes
-		free(sym);
+		struct SymbolListNode *end = purgeList;
+
+		while (end && end->next)
+			end = end->next;
+
+		struct SymbolListNode *node = malloc(sizeof(struct SymbolListNode));
+
+		node->this = sym;
+		node->next = NULL;
+
+		if (end)
+			end->next = node;
+		else
+			purgeList = node;
+	}
+}
+
+// Remove purged symbols
+void sym_FlushPurged(void)
+{
+	struct SymbolListNode *node = purgeList;
+
+	purgeList = NULL;
+	while (node) {
+		struct SymbolListNode *next = node->next;
+		struct Symbol *sym = node->this;
+
+		free(node);
+		node = next;
+
+		if (sym) {
```

Comment:
Will there ever be a case where `sym` can be `NULL` here? If not, don't have an `if`, just `assert(sym)`.